### ISSUE TITLE
I cannot run build and still make auth0 work

### ISSUE THREAD

**Issue by raducretu** [sidebase/nuxt-auth#228]
### Environment

- Operating System: `Darwin`
- Node Version:     `v16.15.1`
- Nuxt Version:     `3.1.1`
- Nitro Version:    `2.1.1`
- Package Manager:  `npm@8.11.0`
- Builder:          `vite`
- User Config:      `css`, `app`, `runtimeConfig`, `modules`, `i18n`, `algolia`, `image`, `auth`
- Runtime Modules:  `@nuxtjs/tailwindcss@6.2.0`, `@nuxtjs/i18n@8.0.0-beta.4-27816734.493bb8f`, `@nuxtjs/algolia@1.2.0`, `nuxt-lodash@2.4.1`, `@sidebase/nuxt-auth@0.4.1`
- Build Modules:    `-`

### Reproduction

setup auth0 as a provider
npm run build
node .output/server/index.mjs 

### Describe the bug

In dev everything works fine, but when i run npm run build and then run de node server it does not work.

### Additional context

nuxt.config.js

```ts
auth: {
      origin: 'http://localhost:3000'
  }
```

[...].ts
```ts
providers: [
        // @ts-ignore
        Auth0Provider.default({
            clientId: config.NUXT_AUTH0_CLIENT_ID,
            clientSecret: config.NUXT_AUTH0_CLIENT_SECRET,
            issuer: config.NUXT_AUTH0_ISSUER
        })
    ]
```
has secret added

### Logs

```shell
[next-auth][error][SIGNIN_OAUTH_ERROR] 
https://next-auth.js.org/errors#signin_oauth_error only valid absolute URLs can be requested {
  error: {
    message: 'only valid absolute URLs can be requested',
    stack: 'TypeError: only valid absolute URLs can be requested\n' +
      '    at Function.request (/Users/raducretu/Work/adn-storefront/.output/server/node_modules/openid-client/lib/helpers/request.js:68:11)\n' +
      '    at Function.discover (/Users/raducretu/Work/adn-storefront/.output/server/node_modules/openid-client/lib/issuer.js:144:38)\n' +
      '    at openidClient (/Users/raducretu/Work/adn-storefront/.output/server/node_modules/next-auth/core/lib/oauth/client.js:16:41)\n' +
      '    at getAuthorizationUrl (/Users/raducretu/Work/adn-storefront/.output/server/node_modules/next-auth/core/lib/oauth/authorization-url.js:70:49)\n' +
      '    at Object.signin (/Users/raducretu/Work/adn-storefront/.output/server/node_modules/next-auth/core/routes/signin.js:38:60)\n' +
      '    at AuthHandler (/Users/raducretu/Work/adn-storefront/.output/server/node_modules/next-auth/core/index.js:253:39)\n' +
      '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +
      '    at async file:///Users/raducretu/Work/adn-storefront/.output/server/chunks/nitro/node-server.mjs:430:24\n' +
      '    at async Object.handler (file:///Users/raducretu/Work/adn-storefront/.output/server/node_modules/h3/dist/index.mjs:840:19)\n' +
      '    at async Server.toNodeHandle (file:///Users/raducretu/Work/adn-storefront/.output/server/node_modules/h3/dist/index.mjs:915:7)',
    name: 'TypeError'
  },
  providerId: 'auth0',
  message: 'only valid absolute URLs can be requested'
}
```

**Comment by raducretu** [comment]
I have managet to debug this further and the problem is on the Auth0Provider.default json, as it does not take env variables in the account, neight by process.env, neither by config. so the way I set it up now to work properly is by chainging all 3 just to be sure :)

`
process.env. config.NUXT_AUTH0_CLIENT_ID || config.NUXT_AUTH0_CLIENT_ID || 'literal client id for production'
`

**Comment by raducretu** [comment]
Still, on Netlify with the netlify preset of Nuxt I still get this error upon login:
<img width="1243" alt="Screenshot 2023-02-13 at 17 11 12" src="https://user-images.githubusercontent.com/85927/218496078-2ae23aab-61b5-48a6-baa2-257e9bac6232.png">

**Comment by zoey-kaiser** [comment]
> I have managed to debug this further and the problem is on the Auth0Provider.default json, as it does not take env variables in the account, neight by process.env, neither by config. so the way I set it up now to work properly is by chainging all 3 just to be sure :)

Hi @raducretu!

I think you are correct in assuming that this is an issue with the environment variables not being passed correctly. As from what I understand, it does work if you hardcode the secrets, correct? 

One approach I could recommend it using runtime environment variables. 

```ts
const runtimeConfig = useRuntimeConfig()

console.log(runtimeConfig.apiSecret)
```

These can either be set in the `nuxt.config.ts` or via the `NUXT_` prefix as normal environment variables. See more here: https://nuxt.com/docs/guide/going-further/runtime-config

**Comment by raducretu** [comment]
in my code config ia actually useRuntingConfig(), just didn’t mention.
neither way works on n run build and then runnibg the node server. Only hard coded values work, which is stragnge as on npm run dev, both work flawlessly.

Also deploying on Netlify the checks parameter of auth0 needs to be set to ‘both’ otherwise it won’t work also

**Comment by zoey-kaiser** [comment]
> in my code config ia actually useRuntingConfig(), just didn’t mention. neither way works on n run build and then runnibg the node server. Only hard coded values work, which is stragnge as on npm run dev, both work flawlessly.
> 
> Also deploying on Netlify the checks parameter of auth0 needs to be set to ‘both’ otherwise it won’t work also

That is super weird 🤔. I don't have any initial ideas of what may be causing this. Have you asked on the offical Nuxt Discord before? Maybe they can help you get some more insights (feel free to share them afterwards here with us as well 🤗)

**Comment by raducretu** [comment]
Anybody any ideea? I tried the discord but nobody answered

**Comment by thdepauw** [comment]
> Still, on Netlify with the netlify preset of Nuxt I still get this error upon login: <img alt="Screenshot 2023-02-13 at 17 11 12" width="1243" src="https://user-images.githubusercontent.com/85927/218496078-2ae23aab-61b5-48a6-baa2-257e9bac6232.png">

Hi,

I'm having the same issue on Netlify using auth0 and `@sidebase/nuxt-auth": "0.4.2`.

I checked if all my environment variables on Netlify are loaded correctly and added a temp api endpoint to check:

```ts
// server\api\health.get.ts
export default defineEventHandler(async (event) => {
  try {
    return { statusCode: 200, body: useRuntimeConfig() };
  } catch (e) {
    console.error(e);
    return createError("Internal Server Error");
  }
});
```

This gave me the same result as running it locally (having the required variables set in the .env file).

Did you by any change find out more about this issue?

**Comment by dpmillerau** [comment]
I'm seeing something similar, but using Azure. I noticed that the error is on the client, and the data is in the private section of runtimeConfig, so the client doesn't see it. And obviously, the client doesn't get AUTH_ORIGIN if you override that.

**Comment by zoey-kaiser** [comment]
Hi @dpmillerau, @thdepauw and @raducretu,

Sorry for being a bit behind this issue. I am currently investigating some other issues that people faced with Netfliy, Vercel and a few other hosters. I hope to gain more insight into their systems and how `nuxt-auth` is best used with them. Once I have more insight ill let you know.

About the environment variables. I managed to deploy the `nuxt-auth-example` on Netlify with the correct environment variables (but then ran into another issue 😵). Maybe you can follow the code we used there?
https://github.com/sidebase/nuxt-auth-example/blob/main/server/api/auth/%5B...%5D.ts

**Comment by raducretu** [comment]
My initial problem started while trying to deploy to Netlify with Auth0 (not credentials) actually and then found out that it does not work to build on local also. The way I made it work on Netlify with Auth0 is by doing the following:

Hard coded the env variables, as they work on dev, but not on build.
Added checks: 'both'

```
const config = useRuntimeConfig()
export default NuxtAuthHandler({
    secret: '--some other secret--',
    debug: true,
    logger:{
        error(code, metadata) {
            console.error(code, metadata)
        },
        warn(code) {
            console.warn(code)
        },
        debug(code, metadata) {
            console.log(code, metadata)
        }
    },
    callbacks: {
        async jwt({token, account, profile}) {
            if (account) {
                token.accessToken = account.access_token
                token.profile = profile
            }
            return token
        },
        async session({session, token: {profile, accessToken}, user}) {

            session.user = session.user || {}
            // @ts-ignore
            const data = await $fetch(`${config.NUXT_AUTH_BASE}/get-user?id=${profile['https://mysite.com/uuid']}&token=token`)
            // @ts-ignore
            session.user = {...session.user, ...data}
            return session
        }
    },
    providers: [
        // @ts-ignore
        Auth0Provider.default({
            checks: 'both',
            clientId: process.env.NUXT_AUTH0_CLIENT_ID || config.NUXT_AUTH0_CLIENT_ID || '--some client id--',
            clientSecret: process.env.NUXT_AUTH0_CLIENT_SECRET || config.NUXT_AUTH0_CLIENT_SECRET || '--some secret--',
            issuer:  process.env.NUXT_AUTH0_ISSUER || config.NUXT_AUTH0_ISSUER || 'https://login.mysite.com'
        })
    ]
})
```

**Comment by smastrom** [comment]
I also had issues with Netlify and spent a couple of hours in vain trying to make things work. While worked as expected on localhost, I started getting callback errors on production.

What can I say is that after trying on Vercel with the same identical Netlify configuration, it worked at first shot. In the meantime I'm now tied to Vercel infrastructure, hoping things with Netlify will sort out.

I'll share my same config maybe it will be helpful.

1. auth0 Dashboard:

- Application URIs / Allowed Callback URLs:

```text
http://localhost:3000/api/auth/callback/auth0,
https://your-repo-name.vercel.app/api/auth/callback/auth0
```

- Application URIs / Allowed Callback URLs:

```text
http://localhost:3000, https://your-repo-name.vercel.app
```

2. .env file at the root of the project on my machine:

```env
NUXT_AUTH_SECRET="REPLACE"
NUXT_AUTH0_CLIENT_ID='REPLACE'
NUXT_AUTH0_CLIENT_SECRET='REPLACE'
NUXT_AUTH0_ISSUER='REPLACE'
NUXT_AUTH0_ORIGIN='http://localhost:3000'
```
- **NUXT_AUTH0_ISSUER** corresponds to the auth0 custom domain, e.g. `https://yourdomain.eu.auth0.com`
- **NUXT_AUTH_SECRET** wraps with **double quotes** a secret you can obtain from the terminal by running:

`node -e "console.log(require('crypto').randomBytes(256).toString('base64'));"` 

3. server/api/auth/[...].ts 
```ts
import { NuxtAuthHandler } from '#auth'
import Auth0 from 'next-auth/providers/auth0'

export default NuxtAuthHandler({
   secret: process.env.NUXT_AUTH_SECRET,
   providers: [
      // @ts-expect-error
      Auth0.default({
         clientId: process.env.NUXT_AUTH0_CLIENT_ID,
         clientSecret: process.env.NUXT_AUTH0_CLIENT_SECRET,
         issuer: process.env.NUXT_AUTH0_ISSUER,
      }),
   ],
})
```
4. nuxt.config.ts

```ts
export default defineNuxtConfig({
   modules: ['@sidebase/nuxt-auth'],
   auth: {
      origin: process.env.NUXT_AUTH0_ORIGIN,
   },
})
```

5. On Vercel when you deploy, simply replace the default command from `nuxi generate` to `pnpm build` (or npm run build, or yarn build according to your package manager) and add the same env variables you have on the machine besides the origin which should be:

NUXT_AUTH0_ORIGIN='http://your-repo-name.vercel.app'.

Hope it helps!

**Comment by zoey-kaiser** [comment]
Wow @smastrom 🤩 

Thank you for that amazing investigation and guide! I think we can determine for now that Netlify is a bit tricky in regards to using `nuxt-auth`. I have tried looking into it myself but have not had much success so far. 

@smastrom is it okay if I run through your vercel guide and add it to our docs (Ill credit you 😄).

I also want to try and close this issue. I know that we could not fix the Netlify issues, but I think it makes sense to start a fresh investigation into it. One of you is welcome to start it, otherwise I will look into it myself and open a new issue to track the progress.

**Comment by smastrom** [comment]
@zoey-kaiser Sure you can! I also just renamed a couple of env vars so that its easier to diff between auth0-specific and nuxt-auth. 😀

**Comment by dpmillerau** [comment]
Ok, I've worked out my issue.

I had code like

   auth: {
        origin: <string>process.env.NUXT_AUTH_ORIGIN || 'http://localhost:3000/',
        enableGlobalAppMiddleware: true
    },

It seems nuxi build does not replace the env variable, OR it doesn't evaluate truthy. Either way, I removed the or clause, and it works ok.


---
David Miller
IT Specialist
First Aid Officer
WHS Officer
Global Recordings Network Australia
Telling the story of Jesus to every people.
Office: (02) 9899-2211 or +61-2-9899-2211
Direct: 02 8014 6481 ext 120
Mobile: 0416-18-3349   or +61-416-18 3349
***@***.***
www.globalrecordings.net/au
Skype dpmillerau
www.facebook.com/davidpmillerau


________________________________

**Comment by thdepauw** [comment]
@zoey-kaiser Can you share the link where you added the information of @smastrom to the Vercel docs? 

I'm trying to deploy to Vercel instead of Netlify, but I'm running into some issues. Was hoping to go through some docs to see if I might have missed something in my setup.

**Comment by dpmillerau** [comment]
BTW I was deploying using Gitlab into a GKE environment.

David Miller
________________________________

**Comment by thdepauw** [comment]
I managed to sort things out and move successfully to Vercel with a working auth0 configuration. 
The example repo on sidebase helped out to debug my problems!

For those interested, it ended up to be an issue with the folder structure in my nuxt project that messed up my deploy on Vercel. I added a folder `api` under `src`. This held some modules with mock data when I didn't have a database connected. It turns out that having this folder messes up the deployment on Vercel and results in having 404 errors on the `api/auth/session` call. After some searching, I noticed none of the `server/api` function were working correctly. The deployment also stated it had found 2 serverless function folders to deploy, while this was only one in the sidebase example.
After removing my own created `api` folder everything was working correctly. 
Note that my `api` folder never created an issue for local development and builds for production run in preview mode.
